c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -193,11 +197,25 @@ struct THCCachingAllocator
     Block* block = NULL;
     Block* remaining = NULL;
 
-    auto it = free_blocks.lower_bound(&search_key);
-    if (it != free_blocks.end() && (*it)->device == device && (*it)->stream == stream) {
-      block = *it;
-      free_blocks.erase(it);
-    } else {
+    for (int i = 0; i < 2; ++i) {
```

Comment:
So, I think there is a straight up problem with this loop as written, which is that if the block is not found, you run `free_blocks.lower_bound` on it twice.

It might be easiest to just put the block search in a lambda and then call it twice explicitly.